src/invariant/InvariantTests.cpp

```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+{
+    auto seed = std::chrono::system_clock::now().time_since_epoch().count();
+    std::default_random_engine gen(seed);
+
+    for (auto t : xdr::xdr_traits<LedgerEntryType>::enum_values())
+    {
+        size_t nTests = 0;
+        while (nTests < 10)
+        {
+            LedgerEntryType let = static_cast<LedgerEntryType>(t);
+
+            VirtualClock clock;
+            Application::pointer appGenerate
+              = Application::create(clock, getTestConfig(0));
+            Application::pointer appApply
+              = Application::create(clock, getTestConfig(1));
+            uint32_t ledgerSeq = generateLedgers(appGenerate, 2, 100);
+
+            // NOTE: We only advance nTests if allLastModified is not empty
+            //       so that we make sure we actually performed a test.
+            std::vector<uint32_t> allLastModified =
+                f(appGenerate->getDatabase(), let, ledgerSeq);
+            if (!allLastModified.empty())
+            {
+                auto dist = std::uniform_int_distribution<size_t>(
+                        0, allLastModified.size()-1);
+                auto index = dist(gen);
+                uint32_t chosenLedger = allLastModified.at(index);
+                REQUIRE_THROWS_AS(
+                  applyBucketsAndCrankUntilDone<T>(
+                      appGenerate, appApply, ledgerSeq,
+                      chosenLedger, let),
+                  InvariantDoesNotHold);
+                ++nTests;
+            }
+        }
+    }
+}
+
+std::vector<uint32_t> getLedgerSeq(Database& db, LedgerEntryType let, uint32_t)
+{
+    std::vector<uint32_t> allLastModified;
+    uint32_t lastModified = 0;
+    auto table = getTableName(let);
+    auto prep = db.getPreparedStatement(
+        "SELECT lastmodified FROM " + table +
+        " WHERE lastmodified > 1");
+    auto& st = prep.statement();
+    st.exchange(soci::into(lastModified));
+    st.define_and_bind();
+    st.execute(true);
+    while (st.got_data())
+    {
+        allLastModified.push_back(lastModified);
+        st.fetch();
+    }
+    return allLastModified;
+}
+
+class ApplyBucketsWorkAddEntry : public ApplyBucketsWork
+{
+  private:
+    uint32_t mFromLedgerSeq;
+    LedgerEntryType mType;
+    bool mAdded;
+
+  public:
+    ApplyBucketsWorkAddEntry(
+            Application& app, WorkParent& parent,
+            std::map<std::string, std::shared_ptr<Bucket>> const& buckets,
+            HistoryArchiveState const& applyState,
+            uint32_t addAtLedgerSeq,
+            LedgerEntryType addType)
+        : ApplyBucketsWork(app, parent, buckets, applyState)
+        , mFromLedgerSeq{addAtLedgerSeq}
+        , mType{addType}
+        , mAdded{false}
+    {
+    }
+
+    Work::State onSuccess() override
+    {
+        if (!mAdded)
+        {
+            auto& db = mApp.getDatabase();
+            size_t count = 0;
+            {
+                auto prep = db.getPreparedStatement(
+                    "SELECT"
```

Comment:
I would rather not have raw SQL code in test cases like this, and instead have those statements in the respective class, for example, here you can use the `*Frame::countObjects` methods - in order to avoid duplication, maybe we just need to have those methods take a min and max, and this test can just have `min==max` ?